### DISCUSSION TITLE
Unable to copy web maps and web applications from Dev to Prod portal

### DISCUSSION THREAD

**Discussion by TejwantSaini** [EsriPS/arcgis-assistant-feedback#24] [Q&A] [answered]
Hi Team,

Background: I have multi-machine ArcGIS Enterprise 10.8.1 installed on Dev and Prod environment using Cloud Builder for Azure.
I am trying to use Assistant to copy portal content - web mapping applications and web maps from dev portal to Production portal. Layers and services are getting copied easily but when I try to copy web maps and web applications it keeps on running without throwing any error. When I investigated the issue using Developer tools. I found the below error in Console tab :

**Access to fetch at 'https://<portal URL>/portal/sharing/rest/content/users/tejwantkaurs@<domain>/c4a0f33b7c474a86a11c13ccbe38ff06/addItem' from origin 'https://assistant.esri-ps.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.**
request.ts:367 POST https://<portal **URL>/portal/sharing/rest/content/users/tejwantkaurs@<domain>/c4a0f33b7c474a86a11c13ccbe38ff06/addItem net::ERR_FAILED**
(anonymous) @ request.ts:367
Promise.then (async)
s @ request.ts:290
(anonymous) @ create.ts:120
Promise.then (async)
u @ create.ts:88
(anonymous) @ add.ts:51
(anonymous) @ copy.ts:159
(anonymous) @ operation.ts:115
(anonymous) @ index.js:5
(anonymous) @ index.js:17
c @ runtime.js:63
(anonymous) @ runtime.js:293
(anonymous) @ runtime.js:118
n @ asyncToGenerator.js:3
l @ asyncToGenerator.js:25
Promise.then (async)
n @ asyncToGenerator.js:13
l @ asyncToGenerator.js:25
(anonymous) @ asyncToGenerator.js:32
(anonymous) @ asyncToGenerator.js:21
(anonymous) @ index.js:7
(anonymous) @ index.js:17
c @ runtime.js:63
(anonymous) @ runtime.js:293
(anonymous) @ runtime.js:118
n @ asyncToGenerator.js:3
l @ asyncToGenerator.js:25
Promise.then (async)
n @ asyncToGenerator.js:13
l @ asyncToGenerator.js:25
(anonymous) @ asyncToGenerator.js:32
(anonymous) @ asyncToGenerator.js:21
(anonymous) @ index.js:7
(anonymous) @ index.js:23
i @ index.js:3
o @ index.js:5
(anonymous) @ operation.ts:127
(anonymous) @ index.js:91
i @ construct.js:19
t @ wrapNativeSuper.js:26
(anonymous) @ createSuper.js:14
n @ index.js:90
M @ operation.ts:105
Fe @ copy.ts:155
(anonymous) @ copy.ts:32
He @ copy.ts:29
(anonymous) @ CopyItems.js:292
c @ runtime.js:63
(anonymous) @ runtime.js:293
(anonymous) @ runtime.js:118
r @ asyncToGenerator.js:3
l @ asyncToGenerator.js:25
(anonymous) @ asyncToGenerator.js:32
(anonymous) @ asyncToGenerator.js:21
(anonymous) @ CopyItems.js:290
(anonymous) @ CopyItems.js:290
(anonymous) @ CopyItems.js:288
action @ CopyItems.js:441
l @ react-dom.production.min.js:14
p @ react-dom.production.min.js:14
(anonymous) @ react-dom.production.min.js:14
v @ react-dom.production.min.js:15
at @ react-dom.production.min.js:52
ot @ react-dom.production.min.js:51
lt @ react-dom.production.min.js:52
pt @ react-dom.production.min.js:56
z @ react-dom.production.min.js:287
H @ react-dom.production.min.js:19
Zt @ react-dom.production.min.js:70
Jt @ react-dom.production.min.js:69
t.unstable_runWithPriority @ scheduler.production.min.js:19
Wo @ react-dom.production.min.js:122
R @ react-dom.production.min.js:287
Kt @ react-dom.production.min.js:68
Show 41 more frames
CopyItems.js:328 TypeError: Failed to fetch
(anonymous) @ CopyItems.js:328
c @ runtime.js:63
(anonymous) @ runtime.js:293
(anonymous) @ runtime.js:118
r @ asyncToGenerator.js:3
u @ asyncToGenerator.js:29
Promise.then (async)
r @ asyncToGenerator.js:13
l @ asyncToGenerator.js:25
(anonymous) @ asyncToGenerator.js:32
(anonymous) @ asyncToGenerator.js:21
(anonymous) @ CopyItems.js:290
(anonymous) @ CopyItems.js:290
(anonymous) @ CopyItems.js:288
action @ CopyItems.js:441
l @ react-dom.production.min.js:14
p @ react-dom.production.min.js:14
(anonymous) @ react-dom.production.min.js:14
v @ react-dom.production.min.js:15
at @ react-dom.production.min.js:52
ot @ react-dom.production.min.js:51
lt @ react-dom.production.min.js:52
pt @ react-dom.production.min.js:56
z @ react-dom.production.min.js:287
H @ react-dom.production.min.js:19
Zt @ react-dom.production.min.js:70
Jt @ react-dom.production.min.js:69
t.unstable_runWithPriority @ scheduler.production.min.js:19
Wo @ react-dom.production.min.js:122
R @ react-dom.production.min.js:287
Kt @ react-dom.production.min.js:68

**CopyItems.js:348 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'params')**
    at CopyItems.js:348
    at Array.map (<anonymous>)
    at CopyItems.js:348
    at Array.filter (<anonymous>)
    at CopyItems.js:346

I checked the Portal Allow Origin setting and it is set to * and when I checked the Network tab there was an error on addItem operation , the details are

Request URL: https://<portalURL>/portal/sharing/rest/content/users/tejwantkaurs@<domain>/addItem
Referrer Policy: strict-origin-when-cross-origin
Connection: keep-alive
Content-Length: 581
Content-Type: text/html
Date: Mon, 13 Sep 2021 05:55:37 GMT
Server: Microsoft-Azure-Application-Gateway/v2

I checked with my IT team and they told me that Web Application Firewall(WAF) is implemented on the Azure network on Prod environment and the request from assistant is using query parameters which vary with every request and are unable to fathom this and they cannot disable WAF due to security implications. Can you please advise what can be causing this and what can I do to make it work?

Regards,
Tejwant Kaur

**Comment by jpeterson** [comment]
Hey @TejwantSaini, sorry for the delay in replying here. Thank you so much for the very detailed report – we will try to reproduce this and get back to you with a workaround or other solution.

cc @nheminger

**Comment by nheminger** [comment]
Hi @TejwantSaini,

Are you able to look at your browser's network tab and paste in the response headers from the addItem call? It appears the server configuration is lacking some of the customary CORS headers.

**Comment by TejwantSaini** [comment, marked as answer]
Hi Nheminger,

Apologies for the delay in response. I checked with my IT team and they opened up some firewall rules on WAF which helped me to use Assistant. They followed the "ArcGIS Enterprise Web Application Filter Rules" document from Esri and also enabled the javascript files on firewall. These changes enabled me to copy maps and apps across from dev to prod using assistant without making any changes to any settings in Portal/server.

Regards,
Tejwant Kaur
